fem/BilinearForm_IBFI.cpp

```diff
@@ -0,0 +1,86 @@
+#include "bilinearform_IBFI.hpp"
+
+using namespace mfem;
+
+BilinearFormIBFI::BilinearFormIBFI(FiniteElementSpace* f) : BilinearForm(f)
+{
+    MFEM_VERIFY(interior_boundary_face_integs_marker != NULL,
+        "interior_boundary_face_integs_marker is NULL.");
+}
+
+void BilinearFormIBFI::AddInteriorBoundaryFaceIntegrator(BilinearFormIntegrator* bfi,
+    Array<int>& int_bdr_marker)
+{
+    interior_boundary_face_integs.Append(bfi);
+    interior_boundary_face_integs_marker.Append(&int_bdr_marker);
+}
+
+void BilinearFormIBFI::Assemble(int skip_zeros) {
+    
+    Mesh* mesh = fes->GetMesh();
+    
+    BilinearForm::Assemble(skip_zeros);
+
+    if (interior_boundary_face_integs.Size())
+    {
+        FaceElementTransformations* tr;
+        Array<int> vdofs2;
+        const FiniteElement* fe1, * fe2;
+
+        // Which interior boundary attributes need to be processed?
+        Array<int> int_bdr_attr_marker(mesh->bdr_attributes.Size() ?
+            mesh->bdr_attributes.Max() : 0);
+        int_bdr_attr_marker = 0;
+        for (int k = 0; k < interior_boundary_face_integs.Size(); k++)
+        {
+            if (interior_boundary_face_integs_marker[k] == NULL)
+            {
+                int_bdr_attr_marker = 1;
+                break;
+            }
+            Array<int>& int_bdr_marker = *interior_boundary_face_integs_marker[k];
+            MFEM_ASSERT(int_bdr_marker.Size() == int_bdr_attr_marker.Size(),
+                "invalid boundary marker for boundary face integrator #"
+                << k << ", counting from zero");
+            for (int i = 0; i < int_bdr_attr_marker.Size(); i++)
+            {
+                int_bdr_attr_marker[i] |= int_bdr_marker[i];
+            }
+        }
+
+        for (int i = 0; i < fes->GetNBE(); i++)
+        {
+            const int mesh_bdr_attr = mesh->GetBdrAttribute(i);
+            if (int_bdr_attr_marker[mesh_bdr_attr - 1] == 0) { continue; }
+
+            tr = mesh->GetInteriorFaceTransformations(i);
```

Comment:
One issue I notice here is that `i` is a boundary element index whereas `GetInteriorFaceTransformations` expects a face index. One can convert `i` to face index with `Mesh::GetBdrFace`, however, that is probably best to be done inside a like `Mesh::GetBdrFaceTransformations` (called e.g. `GetInteriorBdrFaceTransformations`) that initializes the returned object with the proper boundary attribute, boundary element index, etc.

Reply:
We've had an internal discussion about this and we'd like to extend it to you.

The way we originally planned to use this was by marking boundaries in the mesh file with a specific attribute number that later on would be used to pass different internal conditions (such as Total Field, Plane Wave, for example).

Considering the previous paragraph, it is true that GetNBE() returns an index that iterates through the number of boundary elements which also includes our manually assigned 'attributed interior boundaries'. We couldn't help but notice by making a call such as `auto type = fes->GetBdrElementType(i);` inside the for loop and noticing the element type which returned, in a 1D case, the `POINT(0)` type, meaning they are already 'faces'.

I guess also understand the potential difference between a Boundary Element (which I guess we're referring to Computational Boundary Faces?), and Faces themselves (which could extend to Interior and Boundary Faces?), but due to the later call to `GetInteriorFaceTransformations(i)`, we'd purge these computational boundary faces from getting assigned within the `Assemble` while keeping the interior faces doing what we initially wanted.

Computational Boundary Faces don't have a Elem2 while our Interior Boundary Faces do have a Elem2, thus by using mask 31 (by default in `GetInteriorFaceTransformations`), we're kind of assuring we're only processing our designated Interior Boundaries.

Maybe `GetInteriorFaceTransformations()` accepting a Boundary Element Index should trigger an exception?

As a basic summary, by using your already implemented methods, it seems we can obtain the desired result. 

Worry not, I'll still implement your consideration.

Thank you very much for your time,
Alejandro M.

Reply:
@v-dobrev As you've indeed mentioned, there was that crucial mistake between the boundary element index and the 'interior' face index that we were expecting. Amusingly, the few tests we'd prepared had played against us in such a way those index matched, giving us a false positive. A new set of tests have indeed confirmed the fault you mentioned.

I'm working on implementing a fix and improving the tests.

Thank you for your help.
Alejandro M.